key.go

```diff
@@ -0,0 +1,777 @@
+	}
+
+	// If Algorithm is set, it must match the specified key parameters.
+	if k.Algorithm != AlgorithmInvalid {
+		expectedAlg, err := k.deriveAlgorithm()
+		if err != nil {
+			return err
+		}
+
+		if k.Algorithm != expectedAlg {
+			return fmt.Errorf(
+				"found algorithm %q (expected %q)",
+				k.Algorithm.String(),
+				expectedAlg.String(),
+			)
+		}
+	}
+
+	return nil
+}
+
+// MarshalCBOR encodes Key into a COSE_Key object.
+func (k Key) MarshalCBOR() ([]byte, error) {
+	var err error
+
+	if k.KeyType == KeyTypeSymmetric {
+		if k.RawKeyParam, err = encMode.Marshal(k.K); err != nil {
+			return nil, err
+		}
+	} else if k.KeyType == KeyTypeEc2 || k.KeyType == KeyTypeOkp {
+		if k.RawKeyParam, err = encMode.Marshal(k.Curve); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("invalid key type: %q", k.KeyType.String())
+	}
+
+	return encMode.Marshal(k.keyStruct)
+}
+
+// UnmarshalCBOR decodes a COSE_Key object into Key.
+func (k *Key) UnmarshalCBOR(data []byte) error {
+	if err := decMode.Unmarshal(data, &k.keyStruct); err != nil {
+		return err
+	}
+
+	switch k.KeyType {
+	case KeyTypeEc2:
+		if k.RawKeyParam == nil {
+			return errors.New("missing Curve parameter (required for EC2 key type)")
+		}
+
+		if err := decMode.Unmarshal(k.RawKeyParam, &k.Curve); err != nil {
+			return err
+		}
+	case KeyTypeOkp:
+		if k.RawKeyParam == nil {
+			return errors.New("missing Curve parameter (required for OKP key type)")
+		}
+
+		if err := decMode.Unmarshal(k.RawKeyParam, &k.Curve); err != nil {
+			return err
+		}
+	case KeyTypeSymmetric:
+		if k.RawKeyParam == nil {
+			return errors.New("missing K parameter (required for Symmetric key type)")
+		}
+
+		if err := decMode.Unmarshal(k.RawKeyParam, &k.K); err != nil {
+			return err
+		}
+	default:
+		// this should not be reachable as KeyType.UnmarshalCBOR would
+		// result in an error during decMode.Unmarshal() above, if the
+		// value in the data doesn't correspond to one of the above
+		// types.
+		return fmt.Errorf("unexpected key type %q", k.KeyType.String())
+	}
+
+	return k.Validate()
+}
+
+// GetPublic returns a crypto.PublicKey generated using Key's parameters.
+func (k *Key) GetPublic() (crypto.PublicKey, error) {
+	alg, err := k.deriveAlgorithm()
+	if err != nil {
+		return nil, err
+	}
+
+	switch alg {
+	case AlgorithmES256, AlgorithmES384, AlgorithmES512:
+		var curve elliptic.Curve
+
+		switch alg {
+		case AlgorithmES256:
+			curve = elliptic.P256()
+		case AlgorithmES384:
+			curve = elliptic.P384()
+		case AlgorithmES512:
+			curve = elliptic.P521()
+		}
+
+		pub := &ecdsa.PublicKey{Curve: curve, X: new(big.Int), Y: new(big.Int)}
+		pub.X.SetBytes(k.X)
+		pub.Y.SetBytes(k.Y)
+
+		return pub, nil
+	case AlgorithmEd25519:
+		return ed25519.PublicKey(k.X), nil
+	default:
+		return nil, ErrAlgorithmNotSupported
+	}
+}
+
+// GetPrivate returns a crypto.PrivateKey generated using Key's parameters.
+func (k *Key) GetPrivate() (crypto.PrivateKey, error) {
```

Comment:
s/GetPrivate/PrivateKey

PrivateKey is more aligned with the `crypto` naming convention.